File: apps/rust/src/auth/repository.rs
use redis::aio::ConnectionManager;
use redis::AsyncCommands;

use super::session::{SessionConfig, SharedSession};

pub async fn save(
    conn: &mut ConnectionManager,
    config: &SessionConfig,
    session: &SharedSession,
) -> Result<(), redis::RedisError> {
    let key = config.redis_key(&session.session_id);
    let json = serde_json::to_string(session).map_err(|e| {
        redis::RedisError::from((
            redis::ErrorKind::TypeError,
            "session json",
            e.to_string(),
        ))
    })?;
    conn.set_ex(key, json, config.ttl_secs).await
}

pub async fn find_by_id(
    conn: &mut ConnectionManager,
    config: &SessionConfig,
    session_id: &str,
) -> Result<Option<SharedSession>, redis::RedisError> {
    let key = config.redis_key(session_id);
    let json: Option<String> = conn.get(key).await?;
    match json {
        Some(body) if !body.is_empty() => {
            let session: SharedSession = serde_json::from_str(&body).map_err(|e| {
                redis::RedisError::from((
                    redis::ErrorKind::TypeError,
                    "session json",
                    e.to_string(),
                ))
            })?;
            Ok(Some(session))
        }
        _ => Ok(None),
    }
}

pub async fn delete(
    conn: &mut ConnectionManager,
    config: &SessionConfig,
    session_id: &str,
) -> Result<(), redis::RedisError> {
    let key = config.redis_key(session_id);
    conn.del(key).await
}

pub async fn ping(conn: &mut ConnectionManager) -> Result<String, redis::RedisError> {
    redis::cmd("PING").query_async(conn).await
}

File: apps/zig/src/auth/repository.zig
const std = @import("std");
const redis_mod = @import("../redis_client.zig");
const session_mod = @import("session.zig");

pub fn save(
    client: *redis_mod.Client,
    config: *const session_mod.SessionConfig,
    allocator: std.mem.Allocator,
    session: *const session_mod.SharedSession,
) !void {
    const key = try config.redisKey(allocator, session.session_id);
    defer allocator.free(key);
    const json = try session.toJson(allocator);
    defer allocator.free(json);
    try client.setEx(key, json, config.ttl_secs);
}

pub fn findById(
    client: *redis_mod.Client,
    config: *const session_mod.SessionConfig,
    allocator: std.mem.Allocator,
    session_id: []const u8,
) !?session_mod.SharedSession {
    const trimmed = std.mem.trim(u8, session_id, " \t\r\n");
    if (trimmed.len == 0) return null;
    const key = try config.redisKey(allocator, trimmed);
    defer allocator.free(key);
    const raw = try client.get(allocator, key);
    const json = raw orelse return null;
    defer allocator.free(json);
    return try session_mod.SharedSession.parse(allocator, json);
}

pub fn deleteSession(
    client: *redis_mod.Client,
    config: *const session_mod.SessionConfig,
    allocator: std.mem.Allocator,
    session_id: []const u8,
) !void {
    const trimmed = std.mem.trim(u8, session_id, " \t\r\n");
    if (trimmed.len == 0) return;
    const key = try config.redisKey(allocator, trimmed);
    defer allocator.free(key);
    try client.del(key);
}
